File: vnes-emulator/src/main/java/vnes/emulator/ui/GUI.java
package vnes.emulator.ui;
/*
vNES
Copyright © 2006-2013 Open Emulation Project

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program.  If not, see <http://www.gnu.org/licenses/>.
 */


import vnes.emulator.NES;
import vnes.emulator.input.InputHandler;
import vnes.emulator.utils.HiResTimer;

/**
 * UI interface for the NES emulator.
 * This interface defines the core functionality required by any UI implementation,
 * without dependencies on specific UI frameworks like AWT or Compose.
 * It combines both platform-agnostic UI functionality and legacy UI requirements.
 */
public interface GUI {

    // Methods from UiInfoMessageBus
    void showErrorMsg(String message);
    void showLoadProgress(int percentComplete);
    void destroy();

    // GUI-specific methods
    InputHandler getJoy1();
    InputHandler getJoy2();
    ScreenView getScreenView();
    HiResTimer getTimer();
    void imageReady(boolean skipFrame);
    void init(NES nes, boolean showGui);
    void println(String s);
}

File: knes-emulator/src/main/kotlin/knes/emulator/ui/GUI.kt
/*
 *
 *  * Copyright (C) 2025 Artur Skowroński
 *  * This file is part of kNES, a fork of vNES (GPLv3) rewritten in Kotlin.
 *  *
 *  * vNES was originally developed by Brian F. R. (bfirsh) and released under the GPL-3.0 license.
 *  * This project is a reimplementation and extension of that work.
 *  *
 *  * kNES is licensed under the GNU General Public License v3.0.
 *  * See the LICENSE file for more details.
 *
 */

package knes.emulator.ui

import knes.emulator.input.InputHandler
import knes.emulator.utils.HiResTimer

/**
 * UI interface for the knes.emulator.NES emulator.
 * This interface defines the core functionality required by any UI implementation,
 * without dependencies on specific UI frameworks like AWT or Compose.
 * It combines both platform-agnostic UI functionality and legacy UI requirements.
 */

interface GUI {

    // Methods from UiInfoMessageBus
    fun showErrorMsg(message: String)
    fun showLoadProgress(percentComplete: Int)
    fun destroy()

    // GUI-specific methods
    fun getJoy1(): InputHandler
    fun getJoy2(): InputHandler?
    fun getScreenView(): ScreenView
    fun getTimer(): HiResTimer
    fun imageReady(skipFrame: Boolean)
    fun init(papuAppletFunctionality: PAPU_Applet_Functionality, showGui: Boolean)
    fun println(s: String)
}
